### PR TITLE
synchronization fixes for recently discovered deadlock

### PR DESCRIPTION
**Pull request by gabemontero**
Fixes https://issues.jenkins-ci.org/browse/JENKINS-58382

### CODE REVIEW

**Comment by gabemontero** [comment]
@akram @waveywaves PTAL

**Comment by gabemontero** [comment]
@akram - can you and @waveywaves compare based on the instructions at https://github.com/openshift/jenkins-sync-plugin/pull/317#issuecomment-509740769 and find out what addition is needed so we can assign you as a reviewer.

thanks

**Comment by akram** [comment]
@gabemontero I am looking at it

File: src/main/java/io/fabric8/jenkins/openshiftsync/BuildWatcher.java [hunk 1]
```diff
@@ -369,35 +369,37 @@ private static void removeBuildFromNoBCList(Build build) {
...
+                    logger.log(Level.WARNING, "flushBuildsWithNoBCList", t);
+                }
+              }
             }
-          }
-        }
-        if (anyRemoveFailures && buildsWithNoBCList.size() > 0) {
-            buildsWithNoBCList.clear();
+            if (anyRemoveFailures && buildsWithNoBCList.size() > 0) {
+                buildsWithNoBCList.clear();
```
Comment (akram) [inline src/main/java/io/fabric8/jenkins/openshiftsync/BuildWatcher.java]: buildsWithNoBCList  is ConcurrentHashMap, thus:
`(More formally, an update operation for a given key bears a happens-before relation with any (non-null) retrieval for that key reporting the updated value.) For aggregate operations such as putAll and clear, concurrent retrievals may reflect insertion or removal of only some entries.`

Shouldn't we the "clear" synchronised ?

**Comment by akram** [comment]
@gabemontero the Thread Dump attached to the JIRA issue also shows lock on BuildConfigWatcher.onInitialBuildConfigs(BuildConfigWatcher.java:154)  and in io.fabric8.jenkins.openshiftsync.BuildConfigWatcher.upsertJob(BuildConfigWatcher.java:237) .

Is it a side effect of the first deadlock ?

File: src/main/java/io/fabric8/jenkins/openshiftsync/BuildWatcher.java (hunk 1 above)
Comment (gabemontero) [reply src/main/java/io/fabric8/jenkins/openshiftsync/BuildWatcher.java]: ah good catch ... will update @akram

Comment (gabemontero) [reply src/main/java/io/fabric8/jenkins/openshiftsync/BuildWatcher.java]: oh wait ... it is synchronized ... look at line 373:  `synchronized(buildsWithNoBCList) {`

the clean call is in that synchronized block

so we are good here

**Comment by gabemontero** [comment]
Not 100% sure what you are trying to get at with your `BuildConfigWatcher.onInitialBuildConfigs` and `BuildConfigWatcher.upsertJob` comment @akram, but an initial thought which did come to mind is that there are multiple paths/callers to upsertJob .... so synchronizing on the bc's uuid in upsertJob seems better to me.

**Comment by gabemontero** [comment]
@akram @waveywaves help me remember to include discussing this PR in our sync plugin walkthrough today

**Pull request merged**